modules/dataquery/jsx/hooks/usesharedqueries.tsx

```diff
@@ -0,0 +1,417 @@
+import {useState, useEffect} from 'react';
+import swal from 'sweetalert2';
+import {QueryTerm, QueryGroup} from '../querydef';
+import {FlattenedField, FlattenedQuery, VisitOption} from '../types';
+import {
+  APIQuery,
+  APIQueryRun,
+  APIQueryField,
+  APIQueryGroupField,
+  APIQueryCriteriaGroup,
+} from '../types';
+
+type QueryActionType = [
+    (queryID: number|null) => void,
+    (action: string) => void,
+];
+
+interface FlattenedQueryMap {
+    [queryId: number]: FlattenedQuery
+}
+/**
+ * React hook for triggering toggling of starred queries
+ * on a LORIS server.
+ *
+ * @param {function} onCompleteCallback - an action to perform after pinning
+ * @returns {array} - QueryAction functions
+ */
+function useStarredQueries(onCompleteCallback: () => void): QueryActionType {
+    const [starQueryID, setStarQueryID] = useState<number|null>(null);
+    const [starAction, setStarAction] = useState<string>('true');
+    useEffect(() => {
+        if (starQueryID == null) {
+            return;
+        }
+
+        fetch(
+            '/dataquery/queries/' + starQueryID + '?star=' + starAction,
+            {
+                method: 'PATCH',
+                credentials: 'same-origin',
+            },
+        ).then( () => {
+            setStarQueryID(null);
+            if (onCompleteCallback) {
+                onCompleteCallback();
+            }
+        }
+        );
+    }, [starQueryID, starAction]);
+    return [setStarQueryID, setStarAction];
+}
+
+/**
+ * React hook for triggering toggling of shared queries
+ * on a LORIS server.
+ *
+ * @param {function} onCompleteCallback - an action to perform after pinning
+ * @returns {array} - QueryAction functions
+ */
+function useShareQueries(onCompleteCallback: () => void): QueryActionType {
+    const [shareQueryID, setShareQueryID] = useState<number|null>(null);
+    const [shareAction, setShareAction] = useState('true');
+    useEffect(() => {
+        if (shareQueryID == null) {
+            return;
+        }
+
+        fetch(
+            '/dataquery/queries/' + shareQueryID + '?share=' + shareAction,
+            {
+                method: 'PATCH',
+                credentials: 'same-origin',
+            },
+        ).then( () => {
+            setShareQueryID(null);
+            if (onCompleteCallback) {
+                onCompleteCallback();
+            }
+        }
+        );
+    }, [shareQueryID, shareAction]);
+    return [setShareQueryID, setShareAction];
+}
+
+type SharedQueriesType = [
+  {
+        recent: FlattenedQuery[],
+        shared: FlattenedQuery[],
+        top_: FlattenedQuery[],
```

Comment:
wondering if the _ at the end of the key name is necessary / means anything

Reply:
That's a good question. vim was highlighting `top` as if it was a reserved keyword so I added the _.. but I don't think it is? I'll try removing it and see if it works